File: AgoraChat/Code/Controllers/Groups/ACDPublicGroupListViewController.m
//
//  ACDPublicGroupListViewController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/31.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDPublicGroupListViewController.h"
#import "MISScrollPage.h"
#import "AgoraGroupCell.h"
#import "AgoraGroupModel.h"
#import "AgoraNotificationNames.h"
#import "AgoraGroupInfoViewController.h"

#import "ACDGroupNewCell.h"
#import "ACDNoDataPromptView.h"
#import "ACDGroupInfoViewController.h"


#define KPUBLICGROUP_PAGE_COUNT    20

@interface ACDPublicGroupListViewController ()<MISScrollPageControllerContentSubViewControllerDelegate>
@property (nonatomic, strong) ACDNoDataPromptView *noDataPromptView;
@property (nonatomic, strong) NSString *cursor;


@end

@implementation ACDPublicGroupListViewController
#pragma mark life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupNavbar];
    self.view.backgroundColor = UIColor.whiteColor;
    
    [self addNotifications];
    [self loadPublicGroupsFromServer];

    self.table.tableFooterView = [[UIView alloc] init];
    
    ACD_WS
    self.searchResultBlock = ^{
        if (weakSelf.searchResults.count == 0) {
            weakSelf.noDataPromptView.hidden = NO;
        }else {
            weakSelf.noDataPromptView.hidden = YES;
        }
        
        [weakSelf.table reloadData];
    };
    
    self.searchCancelBlock = ^{
        weakSelf.noDataPromptView.hidden = YES;
    };
}

- (void)setupNavbar {
    self.title = @"Public Groups";

    UIButton *leftBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    leftBtn.frame = CGRectMake(0, 0, 20, 20);
    [leftBtn setImage:[UIImage imageNamed:@"gray_goBack"] forState:UIControlStateNormal];
    [leftBtn setImage:[UIImage imageNamed:@"gray_goBack"] forState:UIControlStateHighlighted];
    [leftBtn addTarget:self action:@selector(backAction) forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *leftBar = [[UIBarButtonItem alloc] initWithCustomView:leftBtn];
    [self.navigationItem setLeftBarButtonItem:leftBar];
    
    UIButton *cancelButton = [UIButton buttonWithType:UIButtonTypeCustom];
    cancelButton.frame = CGRectMake(0, 0, 50, 40);
    cancelButton.titleLabel.font = [UIFont systemFontOfSize:16.0];
    [cancelButton setTitleColor:ButtonEnableBlueColor forState:UIControlStateNormal];
    [cancelButton setTitle:@"Cancel" forState:UIControlStateNormal];
    [cancelButton setTitle:@"Cancel" forState:UIControlStateHighlighted];
    [cancelButton addTarget:self action:@selector(backAction) forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *rightBar = [[UIBarButtonItem alloc] initWithCustomView:cancelButton];
    
    UIBarButtonItem *rightSpace = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace
                                                                                target:nil
                                                                                action:nil];
    rightSpace.width = -2;
    [self.navigationItem setRightBarButtonItems:@[rightSpace,rightBar]];
}



- (void)prepare {
    [super prepare];
    [self.view addSubview:self.noDataPromptView];
}

- (void)placeSubViews {
    [super placeSubViews];
    [self.noDataPromptView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.searchBar.mas_bottom).offset(48.0);
        make.centerX.left.right.equalTo(self.view);
    }];
}


- (void)dealloc {
    [self removeNotifications];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
}

- (void)loadPublicGroupsFromServer {
    [self tableViewDidTriggerHeaderRefresh];
}


- (void)addNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(refreshGroupList:) name:KAgora_REFRESH_GROUPLIST_NOTIFICATION object:nil];
}

- (void)removeNotifications {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:KAgora_REFRESH_GROUPLIST_NOTIFICATION object:nil];
}

#pragma mark - Action
- (void)backAction {
    [self.navigationController popViewControllerAnimated:YES];
}

    
#pragma mark - Notification Method
- (void)refreshGroupList:(NSNotification *)notification {
    NSArray *groupList = [[AgoraChatClient sharedClient].groupManager getJoinedGroups];
    [self.dataArray removeAllObjects];
    for (AgoraChatGroup *group in groupList) {
        AgoraGroupModel *model = [[AgoraGroupModel alloc] initWithObject:group];
        if (model) {
            [self.dataArray addObject:model];
        }
    }
    self.searchSource = self.dataArray;
    [self.table reloadData];
}


#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.isSearchState) {
        return [self.searchResults count];
    }
    return self.dataArray.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"AgoraGroupCell";
    ACDGroupNewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell) {
        cell = [[ACDGroupNewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    }
    
    if (self.isSearchState) {
        cell.model = self.searchResults[indexPath.row];
    }else {
        cell.model = self.dataArray[indexPath.row];
    }
    
    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    AgoraGroupModel *model = nil;
    if (self.isSearchState) {
        model = self.searchResults[indexPath.row];
    }else {
        model = self.dataArray[indexPath.row];
    }
    
    if (self.selectedBlock) {
        self.selectedBlock(model.group.groupId);
    }
}


- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 54.0f;
}

#pragma mark - Data
- (void)tableViewDidTriggerHeaderRefresh
{
    self.page = 1;
    [self fetchPublicGroupWithPage:self.page isHeader:YES];
}

- (void)tableViewDidTriggerFooterRefresh
{
    self.page += 1;
    [self fetchPublicGroupWithPage:self.page isHeader:NO];
}

- (void)fetchPublicGroupWithPage:(NSInteger)aPage
                        isHeader:(BOOL)aIsHeader
{
    __weak typeof(self) weakSelf = self;
    if (!aIsHeader) {
        [MBProgressHUD showHUDAddedTo:[UIApplication sharedApplication].keyWindow animated:YES];
    }
    
    [[AgoraChatClient sharedClient].groupManager getPublicGroupsFromServerWithCursor:_cursor pageSize:KPUBLICGROUP_PAGE_COUNT completion:^(AgoraChatCursorResult *aResult, AgoraChatError *aError) {
        [MBProgressHUD hideHUDForView:[UIApplication sharedApplication].keyWindow animated:YES];
//        [self tableViewDidFinishTriggerHeader:YES];
        
        if (!aError) {
            NSArray *groups = [self getGroupsWithResultList:aResult.list];
            if ([_cursor isEqualToString:@""]) {
//                self.showRefreshFooter = NO;
                self.dataArray = [groups mutableCopy];
            }else {
                [self.dataArray addObjectsFromArray:groups];
            }
            weakSelf.cursor = aResult.cursor;
            weakSelf.searchSource = self.dataArray;
            [weakSelf.table reloadData];
        }
    }];
}

- (NSArray *)getGroupsWithResultList:(NSArray *)list {
    NSMutableArray *tGroups = NSMutableArray.new;
    for (AgoraChatGroup *group in list) {
        AgoraGroupModel *model = [[AgoraGroupModel alloc] initWithObject:group];
        if (model) {
            [tGroups addObject:model];
        }
    }
    return [tGroups copy];
}


#pragma mark getter and setter
- (UITableView *)table {
    if (_table == nil) {
        _table                 = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, KScreenHeight) style:UITableViewStylePlain];
        _table.delegate        = self;
        _table.dataSource      = self;
        _table.separatorStyle  = UITableViewCellSeparatorStyleNone;
        _table.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
        _table.clipsToBounds = YES;
    }
    return _table;
}

- (ACDNoDataPromptView *)noDataPromptView {
    if (_noDataPromptView == nil) {
        _noDataPromptView = ACDNoDataPromptView.new;
        [_noDataPromptView.noDataImageView setImage:ImageWithName(@"no_search_result")];
        _noDataPromptView.prompt.text = @"The Group Does Not Exist";
        _noDataPromptView.hidden = YES;
    }
    return _noDataPromptView;
}

#pragma mark getter and setter


@end

#undef KPUBLICGROUP_PAGE_COUNT

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Groups/ACDPublicGroupListViewController.swift
//
//  ACDPublicGroupListViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/20.
//

import UIKit
import AgoraChat
import SnapKit

class ACDPublicGroupListViewController: ACDContainerSearchTableViewController<AgoraGroupModel> {

    private lazy var noDataPromptView: ACDNoDataPlaceHolderView = {
        let view = ACDNoDataPlaceHolderView()
        view.noDataImageView.image = UIImage(named: "no_search_result")
        view.prompt.text = "The Group Does Not Exist".localized
        self.view.addSubview(view)
        view.snp.makeConstraints { make in
            make.top.equalTo(self.searchBar.snp.bottom).offset(48.0)
            make.centerX.left.right.bottom.equalTo(self.view)
        }
        return view
    }()
    
    private var cursor: String?
    
    var selectedHandle: ((_ groupId: String) -> Void)?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "Public Groups".localized
        self.view.backgroundColor = .white
        
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "gray_goBack"), style: .plain, target: self, action: #selector(backAction))
        
        let cancelButton = UIButton(type: .custom)
        cancelButton.frame = CGRect(x: 0, y: 0, width: 50, height: 40)
        cancelButton.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        cancelButton.setTitleColor(Color_154DFE, for: .normal)
        cancelButton.setTitle(LocalizedString.Cancel, for: .normal)
        cancelButton.addTarget(self, action: #selector(backAction), for: .touchUpInside)
        let rightSpace = UIBarButtonItem(barButtonSystemItem: .fixedSpace, target: nil, action: nil)
        rightSpace.width = -2
        self.navigationItem.rightBarButtonItems = [rightSpace, UIBarButtonItem(customView: cancelButton)]
    
        self.table.rowHeight = 54
        
        self.table.separatorStyle = .none
        self.table.keyboardDismissMode = .onDrag
        self.table.tableFooterView = UIView()
        self.table.register(UINib(nibName: "ACDGroupCell", bundle: nil), forCellReuseIdentifier: "cell")
        
        self.searchResultHandle = { [unowned self] in
            self.noDataPromptView.isHidden = self.searchResults.count != 0
            self.table.reloadData()
        }
        self.searchCancelHandle = { [unowned self] in
            self.noDataPromptView.isHidden = true
        }
    
        self.searchBar.snp.remakeConstraints { make in
            make.top.equalTo(self.navigationController?.navigationBar.frame.height ?? 0)
            make.left.right.equalTo(self.view)
        }
        
        NotificationCenter.default.addObserver(self, selector: #selector(refreshGroupList(_:)), name: GroupListChangedNotification, object: nil)
        
        self.fetchPublicGroup(refresh: true)
    }
    
    @objc private func backAction() {
        self.navigationController?.popViewController(animated: true)
    }
    
    private func fetchPublicGroup(refresh: Bool) {
        if !refresh {
            MBProgressHUD.showAdded(to: self.view, animated: true)
        }
        AgoraChatClient.shared().groupManager?.getPublicGroupsFromServer(withCursor: self.cursor, pageSize: 20, completion: { result, error in
            MBProgressHUD.hide(for: self.view, animated: true)
            if let error = error {
                self.showHint(error.errorDescription)
            } else if let list = result?.list {
                let groups = self.getGroupsWithResultList(list)
                if self.cursor == "" {
                    self.dataArray.removeAll()
                }
                self.dataArray.append(contentsOf: groups)
                self.cursor = result?.cursor
                self.searchSource = groups
                self.table.reloadData()
            }
        })
    }

    private func getGroupsWithResultList(_ list: [AgoraChatGroup]) -> [AgoraGroupModel] {
        var groups: [AgoraGroupModel] = []
        for i in list {
            let model = AgoraGroupModel(group: i)
            groups.append(model)
        }
        return groups
    }
    
    @objc private func refreshGroupList(_ notification: Notification) {
        let groupList = AgoraChatClient.shared().groupManager?.getJoinedGroups()
        self.dataArray.removeAll()
        self.searchSource.removeAll()
        if let groupList = groupList {
            for i in groupList {
                let model = AgoraGroupModel(group: i)
                self.dataArray.append(model)
                self.searchSource.append(model)
            }
        }
        self.table.reloadData()
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.isSearchState ? self.searchResults.count : self.dataArray.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDGroupCell {
            if let model = (self.isSearchState ? self.searchResults : self.dataArray)[indexPath.row] as? AgoraGroupModel {
                cell.model = model
            }
        }
        return cell
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let model = (self.isSearchState ? self.searchResults : self.dataArray)[indexPath.row] as? AgoraGroupModel {
            self.selectedHandle?(model.group.groupId)
        }
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}
